front/temporal/tracker/activities.ts

```diff
@@ -0,0 +1,324 @@
+  if (!dataSource) {
+    throw new Error(`Could not find data source ${dataSourceId}`);
+  }
+
+  localLogger.info(
+    "Looking for trackers with watched scope including document."
+  );
+
+  let docParentIds: string[] | null = null;
+
+  if (dataSourceConnectorProvider !== null) {
+    if (dataSource.connectorId === null) {
+      throw new Error(
+        `Data source ${dataSourceId} has no connector but connector provider is set. Cannot find parents.`
+      );
+    }
+
+    const connectorsAPI = new ConnectorsAPI(
+      config.getConnectorsAPIConfig(),
+      logger
+    );
+    const parentsResult = await connectorsAPI.getContentNodesParents({
+      connectorId: dataSource.connectorId,
+      internalIds: [documentId],
+    });
+    if (parentsResult.isErr()) {
+      localLogger.error(
+        {
+          error: parentsResult.error,
+        },
+        "Failed to get parents of document."
+      );
+      return;
+    }
+
+    docParentIds = [
+      documentId,
+      ...parentsResult.value.nodes.flatMap((node) => node.parents),
+    ];
+  }
+
+  const trackers =
+    await TrackerConfigurationResource.internalFetchAllWatchedForDocument(
+      auth,
+      dataSourceId,
+      docParentIds
+    );
+
+  if (!trackers.length) {
+    localLogger.info("No active trackers found for document.");
+    return;
+  }
+
+  const dataSourceDocumentRes = await withRetries(getDataSourceDocument)({
+    workspaceId,
+    dataSourceId,
+    documentId,
+  });
+
+  if (dataSourceDocumentRes.isErr()) {
+    // TODO(DOC_TRACKER): allow to dinstinguish between deleted and "unreachable" docs.
+    localLogger.warn(
+      {
+        error: dataSourceDocumentRes.error,
+      },
+      "Document has been deleted or is unreachable. Skipping post process hook."
+    );
+    return;
+  }
+
+  const dataSourceDocument = dataSourceDocumentRes.value;
+  const documentText = dataSourceDocument.document.text || "";
+  const documentSourceUrl = dataSourceDocument.document.source_url || undefined;
+  if (!documentText) {
+    localLogger.warn(
+      {
+        documentText,
+      },
+      "Document text is empty. Skipping document tracker."
+    );
+    return;
+  }
+
+  // TODO(DOC_TRACKER): Do we skip if source URL is empty?
+  if (!documentSourceUrl) {
+    localLogger.warn(
+      {
+        documentSourceUrl,
+      },
+      "Document source URL is empty. Skipping document tracker."
+    );
+    return;
+  }
+
+  const documentDiff = await getDocumentDiff({
+    dataSource,
+    documentId,
+    hash: documentHash,
+  });
+  const positiveDiff = documentDiff
+    .filter(({ type }) => type === "insert")
+    .map(({ value }) => value)
+    .join("");
+
+  if (
+    positiveDiff.length < TRACKER_WATCHED_DOCUMENT_MINIMUM_POSITIVE_DIFF_LENGTH
+  ) {
+    localLogger.info(
+      { positiveDiffLength: positiveDiff.length },
+      "Positive diff is too short, not running trackers."
+    );
+    return;
+  }
+
+  const diffString = documentDiff
```

Comment:
Not sure I understand this part 🙏🏻

Reply:
this makes a diff string like this basically:
```
- ### old title
+ ### new title
```

We're going to then pass that to the model